test/stages/test_stage_types.py

```diff
@@ -0,0 +1,280 @@
+"""
+Test building stages DAG.
+"""
+
+from pytest_mock import MockFixture
+
+from .. import set_config
+from . import (
+    CohortStage1,
+    DatasetStage1,
+    DatasetStage2,
+    MultiCohortStage1,
+    SGStage1,
+    TestStage,
+    mock_multicohort,
+    mock_multidataset_cohort,
+    run_workflow,
+)
+
+
+def get_config(tmp_path) -> str:
+    conf = f"""
+    [workflow]
+    access_level = 'test'
+    dataset = 'fewgenomes'
+    sequencing_type = 'genome'
+    driver_image = 'stub'
+    input_datasets = ['fewgenomes']
+
+    check_inputs = false
+    check_intermediates = false
+
+    # Skip stages with outputs that already exist
+    check_expected_outputs = false
+
+    [storage.default]
+    default = '{tmp_path}'
+    [storage.fewgenomes]
+    default = '{tmp_path}'
+
+    [hail]
+    billing_project = 'fewgenomes'
+    backend = 'local'
+    dry_run = true
+    """
+
+    return conf
+
+
+def get_config_multicohort(tmp_path) -> str:
+    conf = f"""
+    [workflow]
+    access_level = 'test'
+    dataset = 'fewgenomes'
+    sequencing_type = 'genome'
+    driver_image = 'stub'
+    input_cohorts = ['COHA', 'COHB']
+
+    check_inputs = false
+    check_intermediates = false
+
+    # Skip stages with outputs that already exist
+    check_expected_outputs = false
+
+    [storage.default]
+    default = '{tmp_path}'
+    [storage.fewgenomes]
+    default = '{tmp_path}'
+
+    [hail]
+    billing_project = 'fewgenomes'
+    backend = 'local'
+    dry_run = true
+    """
+
+    return conf
+
+
+def test_dataset_stages(mocker: MockFixture, tmp_path):
+    """
+    Test that dataset stages are run once per dataset.
+    """
+    set_config(get_config(tmp_path=tmp_path), tmp_path / 'config.toml')
+
+    from cpg_utils.hail_batch import get_batch
+    from cpg_workflows.inputs import get_inputs
+
+    run_workflow(mocker, cohort_mocker=mock_multidataset_cohort, stages=[DatasetStage1])
+    print('Job by stage:', get_batch().job_by_stage)
+    assert get_batch().job_by_stage.get('DatasetStage1')
+    # The next two tests are looking at the same behaviour, but from different angles
+    assert get_batch().job_by_stage.get('DatasetStage1').get('job_n') == 2
+    assert get_batch().job_by_stage.get('DatasetStage1').get('job_n') == len(get_inputs().get_datasets())
```

Comment:
First assert that the len(get_inputs().get_datasets())) == 2, then leave 93 as is.